src/shared/fs/templateRegistry.ts

```diff
@@ -72,59 +71,48 @@ export class AsyncCloudFormationTemplateRegistry {
     ) {}
 
     /**
-     * Returns the initial registry instance if setup has completed, otherwise
-     * returning a temporary instance and showing a progress message to the user
-     * to indicate setup is in progress.
+     * Returns the initial registry instance if setup has completed, otherwise returns a temporary
+     * instance and shows a progress message ("Scanning...") until setup is done.
      */
     async getInstance(): Promise<CloudFormationTemplateRegistry> {
         if (this.isSetup) {
             return this.instance
         }
+        if (this.setupPromise) {
+            getLogger().debug('%s: getInstance() requested, still initializing', this.instance.name)
+            return this.setupPromise
+        }
 
-        let perf: PerfLog
+        // Show a "Scanning..." progress message until setup is done.
         const cancelSetup = new Timeout(30 * 60 * 1000) // 30 min
-        if (!this.setupPromise) {
-            perf = new PerfLog('cfn: template registry setup')
-            this.setupPromise = this.asyncSetupFunc(this.instance, cancelSetup)
-        }
+        const msg = localize(
+            'AWS.codelens.waitingForTemplateRegistry',
+            'Scanning CloudFormation templates (except [search.exclude](command:workbench.action.openSettings?"@id:search.exclude") paths)...'
+        )
+        const progress = await showMessageWithCancel(msg, cancelSetup)
+
+        const perf = new PerfLog(`${this.instance.name}: template registry setup`)
+        this.setupPromise = this.asyncSetupFunc(this.instance, cancelSetup, item => {
+            if (cancelSetup.completed) {
+                getLogger().debug('%s: getInstance() cancelled', this.instance.name)
+                return
+            }
+            progress.report({ message: item.toString() })
+        })
+
         this.setupPromise.then(() => {
             if (perf) {
                 perf.done()
             }
             this.isSetup = true
             cancelSetup.dispose()
         })
-        // Show user a message indicating setup is in progress
-        if (this.setupProgressMessage === undefined) {
-            // TODO: use showMessageWithCancel() ?
-            this.setupProgressMessage = vscode.window.withProgress(
-                {
-                    location: vscode.ProgressLocation.Notification,
-                    title: localize(
-                        'AWS.codelens.waitingForTemplateRegistry',
-                        'Scanning CloudFormation templates (except [search.exclude](command:workbench.action.openSettings?"@id:search.exclude") paths)...'
-                    ),
-                    cancellable: true,
-                },
-                async (progress, token) => {
-                    token.onCancellationRequested(() => {
-                        // Allows for new message to be created if templateRegistry variable attempted to be used again
-                        this.setupProgressMessage = undefined
```

Comment:
Don't think we supported this, at least after the Cancel timeout was introduced.